crates/c-api/src/wasi.rs

```diff
@@ -149,6 +154,121 @@ pub extern "C" fn wasi_config_inherit_stdout(config: &mut wasi_config_t) {
     config.builder.inherit_stdout();
 }
 
+struct CustomOutputStreamInner {
+    foreign_data: crate::ForeignData,
+    callback: extern "C" fn(*mut c_void, *const u8, usize) -> isize,
+}
+
+impl CustomOutputStreamInner {
+    pub fn write(&self, buf: &[u8]) -> isize {
+        (self.callback)(self.foreign_data.data, buf.as_ptr(), buf.len())
+    }
+}
+
+pub struct CustomOutputStream {
+    inner: std::sync::Arc<CustomOutputStreamInner>,
+}
+
+impl CustomOutputStream {
+    pub fn new(
+        foreign_data: crate::ForeignData,
+        callback: extern "C" fn(*mut c_void, *const u8, usize) -> isize,
+    ) -> Self {
+        Self {
+            inner: std::sync::Arc::new(CustomOutputStreamInner {
+                foreign_data,
+                callback,
+            }),
+        }
+    }
+}
+
+impl Clone for CustomOutputStream {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl wasmtime_wasi::p2::Pollable for CustomOutputStream {
+    async fn ready(&mut self) {}
+}
+
+#[async_trait::async_trait]
+impl wasmtime_wasi::p2::OutputStream for CustomOutputStream {
+    fn write(&mut self, bytes: Bytes) -> Result<(), StreamError> {
+        let wrote = self.inner.write(&bytes);
+
+        if wrote < 0 {
+            return Err(StreamError::Trap(
+                io::Error::from_raw_os_error(wrote.abs() as _).into(),
+            ));
+        }
+
+        if wrote as usize != bytes.len() {
+            return Err(StreamError::Trap(anyhow::anyhow!(
+                "Partial writes in wasip2 implementation are not allowed"
+            )));
+        }
+
+        Ok(())
+    }
+    fn flush(&mut self) -> Result<(), StreamError> {
+        Ok(())
+    }
+    fn check_write(&mut self) -> Result<usize, StreamError> {
+        Ok(usize::MAX)
+    }
+}
+
+impl AsyncWrite for CustomOutputStream {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        let wrote = self.inner.write(buf);
+        Poll::Ready(if wrote >= 0 {
+            Ok(wrote as _)
+        } else {
+            Err(io::Error::from_raw_os_error(wrote.abs() as _))
+        })
```

Comment:
Mind deduplicating this with the above by having a helper `fn raw_write(&self, bytes: &[u8]) -> io::Result<usize>` or similar?